Add signing request paramenters

skobyda opened the pull request:
Builds on top of #3 

@andreasn @garrett
Allows user to specify additional signing request parameters:
![Screenshot from 2020-06-01 14-51-03](https://user-images.githubusercontent.com/42733240/83410989-ace34500-a417-11ea-938e-93fac19fc02b.png)

![Screenshot from 2020-06-01 14-50-50](https://user-images.githubusercontent.com/42733240/83410985-ac4aae80-a417-11ea-91a6-baa0e87444b1.png)

andreasn commented on src/certificateList.jsx:
```diff
@@ -90,44 +93,81 @@ function getCAName(cas, cert) {
+                <div className='ct-form'>
+                    {cert.subject && cert.subject.v && <>
+                        <label className='control-label label-title' htmlFor={`${idPrefix}-general-subject`}>
+                            {_("Subject Name")}
```
Patternfly4 (unlike Patternfly3) recommends Sentence style
https://www.patternfly.org/v4/design-guidelines/content/grammar-and-terminology#sentence-style

So this would become _Subject name_

andreasn commented on src/certificateList.jsx:
```diff
@@ -90,44 +93,81 @@ function getCAName(cas, cert) {
+
+                    {cert.principal && cert.principal.v.length > 0 && <>
+                        <label className='control-label label-title' htmlFor={`${idPrefix}-general-principal`}>
+                            {_("Principal Name")}
```
Should be Sentence style rather than Headline Style

andreasn commented on src/certificateList.jsx:
```diff
@@ -90,44 +93,81 @@ function getCAName(cas, cert) {
+                <div className='ct-form'>
+                    {cert.hostname && cert.hostname.v.length > 0 && <>
+                        <label className='control-label label-title' htmlFor={`${idPrefix}-general-dns`}>
+                            {_("DNS Name")}
```
Should be Sentence style rather than Headline Style

andreasn commented on src/requestCertificate.jsx:
```diff
@@ -0,0 +1,354 @@
+                type='checkbox'
+                checked={dialogValues.signingParameters}
+                onChange={e => onValueChanged("signingParameters", !dialogValues.signingParameters)} />
+            {_("Set Signing Request Parameters")}
```
Should be Sentence style rather than Headline Style

andreasn commented on src/requestCertificate.jsx:
```diff
@@ -0,0 +1,354 @@
+    return (
+        <>
+            <label className="control-label" htmlFor="subject-name">
+                {_("Subject Name")}
```
Should be Sentence style rather than Headline Style

andreasn commented on src/requestCertificate.jsx:
```diff
@@ -0,0 +1,354 @@
+    return (
+        <>
+            <label className="control-label" htmlFor="principal-name">
+                {_("Principal Name")}
```
Should be Sentence style rather than Headline Style

skobyda replied:
Fixed this and all above

andreasn commented:
I'm also thinking it would be good to clarify the hierarchy between the checkbox and the items that are triggered by the checkbox. So something like this:

![83410989-ace34500-a417-11ea-938e-93fac19fc02b](https://user-images.githubusercontent.com/209895/83427641-5551d300-a431-11ea-8221-f669770cbcf2.png)

garrett commented:
1. Is "Nickname" optional?
   a. If optional, then is a nickname autogenerated? (and if so, can we show the autogenerated one?)
   b. If not optional, is "Nickname" prominently displayed in the list and the main way to tell the difference between certificates? (If so, it should _probably_ be first.)

skobyda commented:
> 1. Is "Nickname" optional?
>    
>    1. If optional, then is a nickname autogenerated? (and if so, can we show the autogenerated one?)
>    2. If not optional, is "Nickname" prominently displayed in the list and the main way to tell the difference between certificates? (If so, it should _probably_ be first.)
> 2. Should "Nickname" be called something else? Nickname is usually reserved for a person, not a thing. (But perhaps it's clear enough here? shrug)

The nickname field is implemented in #3. I put there my idea of autogenerating Nickname (it's compulsory for NSSDB stored certificate).

skobyda commented:
@andreasn I fixed hierarchy and letter case and also changed "Set signing request parameters" to "Set optional..." to perhaps clarify that every one of those fields is optional.

@garrett I also added autogeneration for Subject name which consists of "CN=$hostname"

![Screenshot from 2020-06-03 12-25-05](https://user-images.githubusercontent.com/42733240/83626149-4edc6c00-a595-11ea-9a74-c8c8964cf768.png)

skobyda commented:
So I changed DNS names row to support multiple dns names. Separated by commas. If user tries to enter any whitespace, it's replaced by comma.
@garrett The helper text was a bity wide, spacy so I changed it's line-height to 1.5rem, and and set it's top padding to 0, but there is still bit too space between TextArea and helper text. WDYT?
![Screenshot from 2020-06-04 11-18-44](https://user-images.githubusercontent.com/42733240/83738927-35025e00-a655-11ea-94e0-d29ff8bc936a.png)

garrett commented:
I haven't been able to compile this, so I can't really tweak things to stack the labels differently to compare how it'd look one way versus the other. Sorry.

garrett commented:
Ok, it compiles now, but I get this:

![Screenshot_2020-06-29 Certificates - garrett Rain](https://user-images.githubusercontent.com/10246/86002707-02a80e80-ba11-11ea-87e7-151043633820.png)

Do I need to overlay a specific RPM for this to work? (Like some certmanager or certd package?)

andreasn commented:
I have certmonger overlayed, I think that should do the trick.
I found it most straight forward to test this in a VM.

garrett commented:
>  I found it most straight forward to test this in a VM.

How do you make a VM with cockpit-cert inside of it?

skobyda commented:
> Ok, it compiles now, but I get this:
> 
> ![Screenshot_2020-06-29 Certificates - garrett Rain](https://user-images.githubusercontent.com/10246/86002707-02a80e80-ba11-11ea-87e7-151043633820.png)
> 
> Do I need to overlay a specific RPM for this to work? (Like some certmanager or certd package?)

Made clean fedora32 VM, and yeah it seems you need $sudo dnf install certmonger

skobyda commented:
> > I found it most straight forward to test this in a VM.
> 
> How do you make a VM with cockpit-cert inside of it?

Here are the steps which I had to do to make this repo running on clean fedora32 VM. Will put it into README. Beware that certmonger allows only root to manage certificates.

```
(Install cockpit according to https://cockpit-project.org/running.html)

# git clone https://github.com/skobyda/cockpit-certificates.git
# cd cockpit-certificates
# dnf install certmonger
# dnf install npm
# make
# mkdir -p ~/.local/share/cockpit
# ln -s `pwd`/dist ~/.local/share/cockpit/starter-kit
```

garrett commented:
I guess the README should be updated for developers and we should make sure certmonger is a dep for the package (we probably already do).

garrett commented:
FWIW: I've installed certmonger on my system, but haven't had a chance to reboot yet today. (I'll do so tomorrow. It's late now.)

skobyda commented:
> FWIW: I've installed certmonger on my system, but haven't had a chance to reboot yet today. (I'll do so tomorrow. It's late now.)

@garrett Could take a look?

garrett commented:
Thanks for pinging me about this.

I can't inspect this with devtools, however, as my browser keeps hanging when specifically trying to look at the HTML/JS/CSS of this dialog... It's quite strange. I've tried everything short of a reboot.

garrett commented:
I've rebooted... and same issue with Firefox in Flatpak, Firefox nightly tarball from Mozilla, and Chrome (Google RPM).

Are you doing anything special with JavaScript in this dialog? It freezes every browser. It's like it's caught in some CPU intensive loop. I'm specifically looking at things when the signing request parameters expanded.

garrett commented:
I did see some other issues when looking a the console... like a `<form>` within a `<form>` (that's not valid), and a few other problems.

Also, why are the optional values a part of nickname? That seems... very odd?

Additionally, the version of `lib/form-layout.scss` is not in sync with Cockpit.

garrett commented:
So, please:

- [ ] Sync `lib/form-layout.scss` from Cockpit
- [ ] Fix weird nesting of optional values out from under nickname
- [ ] Fix the nested form issue
- [ ] Look at the error messages in dev tool's console and fix them
- [ ] Debug what's chowing down on CPU and fix the horrible performance issue causing browsers to freeze and crash

And then we can proceed.

Additionally, there are way too many `<hr>`s in this form for some reason. They're supposed to be group dividers, not element dividers. The `grid-gap` in `ct-form` spaces the elements and doesn't need an `<hr>` everywhere.

skobyda commented:
> So, please:
> 
> * [ ]  Sync `lib/form-layout.scss` from Cockpit

Is it not synced? I tried to copy pkg/lib/form-layout.scss from cockpit repo to lib/form-layout.scss in this repo and $git diff tells me there is no difference.

> * [ ]  Fix weird nesting of optional values out from under nickname

The nesting is done under checkbox "Set optional signing request parameters". Perhaps the checkbox should be moved to the left.

> * [ ]  Fix the nested form issue

Fixed
 
> * [ ]  Look at the error messages in dev tool's console and fix them

There were 2 errors related to this PR I fixed. Others inherited from cockpit as they are the same errors as in any other cockpit page.

> * [ ]  Debug what's chowing down on CPU and fix the horrible performance issue causing browsers to freeze and crash

I checked and it's the problem of the dialog in general, not just what I do in this PR. So I created a separate issue for that:
#24 

> And then we can proceed.
> 
> Additionally, there are way too many `<hr>`s in this form for some reason. They're supposed to be group dividers, not element dividers. The `grid-gap` in `ct-form` spaces the elements and doesn't need an `<hr>` everywhere.

Fixed. Removed several <hr>

> (I've tried fixing some of this myself, but I just _cannot_ with that horrible CPU eating bug. Unfortunately, I couldn't find what bit of code makes browsers freeze and crash.)

skobyda commented on lib/form-layout.scss:
```diff
@@ -90,6 +90,12 @@
     font-size: var(--pf-global--FontSize--sm);
   }
 
+  > .pf-c-form__helper-text {
```
@garrett Here I add a fix for spacing for helper text in ct-form

garrett replied:
Why isn't it an alias for `.help-block` at https://github.com/skobyda/cockpit-certificates/blob/0a1a4d45a80601c3a476d651fd8e8df8d4b2ab48/lib/form-layout.scss#L290-L320 ?

skobyda replied:
You mean doing something like this?
```
  > .pf-c-form__helper-text {
    @extend .help-block;
  }
```

skobyda replied:
Fixed it

skobyda commented:
@garrett I fixed most of the stuff. I believe form-layout.scss is up to date as git diff tells me no difference when I copy that file from cockpit repo, I created an issue for CPU problem (it seems it's problem on whole certificates page) and about nesting of optional values that's done according to andreas comment about showing hierarchy between checkbox and the inputs.

garrett commented:
Nice. Building now. Hopefully the performance will be acceptable enough to look at it. (It wasn't last time, oddly.) Perhaps there's some JS that isn't being throttled and/or it's supposed to be called on an event, but instead it's called continually?

skobyda commented:
> Nice. Building now. Hopefully the performance will be acceptable enough to look at it. (It wasn't last time, oddly.) Perhaps there's some JS that isn't being throttled and/or it's supposed to be called on an event, but instead it's called continually?

The perfomance isn't better probably thou, as I haven't had time to look into that, just created a seperate PR since it's problem on a whole page so it doesn't block this PR
